### PR TITLE
Support scale subresource for PDBs

### PR DESCRIPTION
**Pull request by mortent**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
PDBs currently only work with the workload controllers deployment, replicaset, statefulset and replicationcontroller. This change adds support for any controllers that implement the scale subresource. 

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
PDB support for custom resources with scale subresource
```

### CODE REVIEW

**Comment by liggitt** [comment]
thanks for the PR... is there a proposal for this API change?

**Comment by mortent** [comment]
@liggitt I have discussed it with @kow3ns, but there is no KEP for this. I thought this was small enough to not require a KEP, but I'm happy to do one if needed.

**Comment by liggitt** [comment]
Yes, a design would be good. It can be short if the semantics are simple and there are few issues.

**Comment by janetkuo** [comment]
Link to KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/20190412-pdb-support-for-custom-resources-with-scale-subresource.md

Suggest changing release note to: `PDB support for custom resources with scale subresource`

File: pkg/controller/disruption/disruption.go [hunk 1]
```diff
@@ -279,6 +271,34 @@ func (dc *DisruptionController) getPodReplicationController(pod *v1.Pod) (*contr
...
+	gr := mapping.Resource.GroupResource()
+
+	scale, err := dc.scaleNamespacer.Scales(namespace).Get(gr, controllerRef.Name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
```
Comment (janetkuo) [inline pkg/controller/disruption/disruption.go]: Do we need to check scalable resource UID against controllerRef UID, like other `podControllerFinder` functions?

Comment (mortent) [reply pkg/controller/disruption/disruption.go]: Added the check for the scale function too.

**Review by janetkuo — APPROVED** [review]

**Comment by mortent** [comment]
@mikedanese @liggitt Can you take a look at this one? I would like to get this into 1.15 and I don't want to get too close to code freeze.

File: pkg/controller/disruption/disruption.go [hunk 2]
```diff
@@ -279,6 +271,36 @@ func (dc *DisruptionController) getPodReplicationController(pod *v1.Pod) (*contr
...
+		Kind:  controllerRef.Kind,
+	}
+
+	mapping, err := dc.mapper.RESTMapping(gk, gv.Version)
+	if err != nil {
+		return nil, err
+	}
+	gr := mapping.Resource.GroupResource()
+
+	scale, err := dc.scaleNamespacer.Scales(namespace).Get(gr, controllerRef.Name)
```
Comment (liggitt) [inline pkg/controller/disruption/disruption.go]: this API is unfortunate... we're throwing away the version information we just got from our restmapper. if we know the full group/version/resource we want, we should be able to invoke it directly, and the scale client would be way simpler (it wouldn't need a restmapper internally). can you open a follow up issue to improve this?

File: pkg/controller/disruption/disruption.go [hunk 3]
```diff
@@ -583,10 +605,23 @@ func (dc *DisruptionController) getExpectedScale(pdb *policy.PodDisruptionBudget
 	// 1. Find the controller for each pod.  If any pod has 0 controllers,
 	// that's an error. With ControllerRef, a pod can only have 1 controller.
 	for _, pod := range pods {
+		controllerRef := metav1.GetControllerOf(pod)
+		if controllerRef == nil {
+			err = fmt.Errorf("found no controller ref for pod %q", pod.Name)
```
Comment (liggitt) [inline pkg/controller/disruption/disruption.go]: a nil controller didn't log an error previously, right?

File: pkg/controller/disruption/disruption.go [hunk 4]
```diff
@@ -180,15 +188,11 @@ var (
 )
 
 // getPodReplicaSet finds a replicaset which has no matching deployments.
-func (dc *DisruptionController) getPodReplicaSet(pod *v1.Pod) (*controllerAndScale, error) {
-	controllerRef := metav1.GetControllerOf(pod)
-	if controllerRef == nil {
-		return nil, nil
-	}
+func (dc *DisruptionController) getPodReplicaSet(controllerRef *metav1.OwnerReference, namespace string) (*controllerAndScale, error) {
 	if controllerRef.Kind != controllerKindRS.Kind {
```
Comment (liggitt) [inline pkg/controller/disruption/disruption.go]: need to also make sure `schema.ParseGroupVersion(controllerRef.APIVersion).Group == ""`

File: pkg/controller/disruption/disruption.go [hunk 5]
```diff
@@ -204,16 +208,12 @@ func (dc *DisruptionController) getPodReplicaSet(pod *v1.Pod) (*controllerAndSca
...
 
-// getPodStatefulSet returns the statefulset managing the given pod.
-func (dc *DisruptionController) getPodStatefulSet(pod *v1.Pod) (*controllerAndScale, error) {
-	controllerRef := metav1.GetControllerOf(pod)
-	if controllerRef == nil {
-		return nil, nil
-	}
+// getPodStatefulSet returns the statefulset referenced by the provided controllerRef.
+func (dc *DisruptionController) getPodStatefulSet(controllerRef *metav1.OwnerReference, namespace string) (*controllerAndScale, error) {
 	if controllerRef.Kind != controllerKindSS.Kind {
```
Comment (liggitt) [inline pkg/controller/disruption/disruption.go]: need to also make sure `schema.ParseGroupVersion(controllerRef.APIVersion).Group == "apps"`

File: pkg/controller/disruption/disruption.go [hunk 6]
```diff
@@ -226,15 +226,11 @@ func (dc *DisruptionController) getPodStatefulSet(pod *v1.Pod) (*controllerAndSc
 }
 
 // getPodDeployments finds deployments for any replicasets which are being managed by deployments.
-func (dc *DisruptionController) getPodDeployment(pod *v1.Pod) (*controllerAndScale, error) {
-	controllerRef := metav1.GetControllerOf(pod)
-	if controllerRef == nil {
-		return nil, nil
-	}
+func (dc *DisruptionController) getPodDeployment(controllerRef *metav1.OwnerReference, namespace string) (*controllerAndScale, error) {
 	if controllerRef.Kind != controllerKindRS.Kind {
```
Comment (liggitt) [inline pkg/controller/disruption/disruption.go]: need to also make sure `schema.ParseGroupVersion(controllerRef.APIVersion).Group == "apps" or "extensions"`

File: pkg/controller/disruption/disruption.go [hunk 7]
```diff
@@ -260,15 +256,11 @@ func (dc *DisruptionController) getPodDeployment(pod *v1.Pod) (*controllerAndSca
 	return &controllerAndScale{deployment.UID, *(deployment.Spec.Replicas)}, nil
 }
 
-func (dc *DisruptionController) getPodReplicationController(pod *v1.Pod) (*controllerAndScale, error) {
-	controllerRef := metav1.GetControllerOf(pod)
-	if controllerRef == nil {
-		return nil, nil
-	}
+func (dc *DisruptionController) getPodReplicationController(controllerRef *metav1.OwnerReference, namespace string) (*controllerAndScale, error) {
 	if controllerRef.Kind != controllerKindRC.Kind {
```
Comment (liggitt) [inline pkg/controller/disruption/disruption.go]: need to also make sure `schema.ParseGroupVersion(controllerRef.APIVersion).Group == ""`

**Review by liggitt — COMMENTED** [review]
RBAC change looks fine, there's some additional checking that needs to be done on the controllerRef group

File: pkg/controller/disruption/disruption.go (hunk 3 above)
Comment (mortent) [reply pkg/controller/disruption/disruption.go]: A pod without a controller was considered an error before this change. We just happened to check all the finder functions before returning the error. This just bails out early if there is no controllerRef.
https://github.com/kubernetes/kubernetes/blob/25b0d2dbf4521bd1e7bae471d85c369eb89f9555/pkg/controller/disruption/disruption.go#L599-L603

File: pkg/controller/disruption/disruption.go (hunk 2 above)
Comment (mortent) [reply pkg/controller/disruption/disruption.go]: Created issue: https://github.com/kubernetes/kubernetes/issues/78231

File: pkg/controller/disruption/disruption.go (hunk 4 above)
Comment (mortent) [reply pkg/controller/disruption/disruption.go]: Added check for group on all finder functions.

File: pkg/controller/disruption/disruption.go (hunk 3 above)
Comment (liggitt) [reply pkg/controller/disruption/disruption.go]: ah, got it, thanks.

File: pkg/controller/disruption/disruption.go [hunk 8]
```diff
@@ -227,8 +229,9 @@ func (dc *DisruptionController) getPodStatefulSet(controllerRef *metav1.OwnerRef
...
 // getPodDeployments finds deployments for any replicasets which are being managed by deployments.
 func (dc *DisruptionController) getPodDeployment(controllerRef *metav1.OwnerReference, namespace string) (*controllerAndScale, error) {
-	if controllerRef.Kind != controllerKindRS.Kind {
-		return nil, nil
+	ok, err := verifyGroupKind(controllerRef, controllerKindRS.Kind, []string{"apps", "extensions"})
+	if !ok || err != nil {
+		return nil, err
 	}
 	rs, err := dc.rsLister.ReplicaSets(namespace).Get(controllerRef.Name)
 	if err != nil {
```
Comment (liggitt) [inline pkg/controller/disruption/disruption.go]: pre-existing, but the same group check ("apps" or "extensions") should be done when resolving the deployment controllerref below. up to you whether you do that here or in a follow-up

File: pkg/controller/disruption/disruption.go [hunk 9]
```diff
@@ -301,6 +305,29 @@ func (dc *DisruptionController) getScaleController(controllerRef *metav1.OwnerRe
...
+		if group == gv.Group {
+			groupMatch = true
+			break
+		}
+	}
+	if !groupMatch {
+		return false, nil
+	}
+
+	if controllerRef.Kind == expectedKind {
```
Comment (liggitt) [inline pkg/controller/disruption/disruption.go]: nit: move this first and short-circuit the parse and group search and return `false, nil` early if this doesn't match

File: pkg/controller/disruption/disruption.go [hunk 10]
```diff
@@ -301,6 +305,29 @@ func (dc *DisruptionController) getScaleController(controllerRef *metav1.OwnerRe
...
+func verifyGroupKind(controllerRef *metav1.OwnerReference, expectedKind string, expectedGroups []string) (bool, error) {
+	gv, err := schema.ParseGroupVersion(controllerRef.APIVersion)
+	if err != nil {
+		return false, err
+	}
+
+	groupMatch := false
+	for _, group := range expectedGroups {
+		if group == gv.Group {
+			groupMatch = true
```
Comment (liggitt) [inline pkg/controller/disruption/disruption.go]: if the kind match has already been done, can simply `return true, nil` here

**Comment by liggitt** [comment]
a couple nits in verifyGroupKind, and one other place a group check is needed

I'm not that familiar with the test coverage of PDBs... do we have positive and negative tests for each of the controllerref types modified here?

File: pkg/controller/disruption/disruption.go (hunk 8 above)
Comment (mortent) [reply pkg/controller/disruption/disruption.go]: Updated the PR to include the group test on the deployment controller ref.

File: pkg/controller/disruption/disruption.go (hunk 9 above)
Comment (mortent) [reply pkg/controller/disruption/disruption.go]: Yeah, this looks cleaner. I have updated the function.

**Comment by mortent** [comment]
@liggitt I have added additional tests to cover the finder functions.

**Comment by liggitt** [comment]
thanks, looks great

/lgtm
/approve

**Pull request merged**